### ISSUE TITLE
Image cache not updated on file change

### ISSUE THREAD

**Issue by HiFiPhile** [nextcloud/server#17940]
### Steps to reproduce
I'm using rrdtool to log and generate server status graphs every 5 minutes. While the downloaded image is correct, the thumbnail and photo viewer show outdated image. 
1. Use photo viewer to view an image.
2. Replace the image with a similar one (same size)
3. (Optional) Clear browser cache & relogin to narrow the problem
4. Use photo viewer to view the new image

### Expected behaviour
New image will be shown.

### Actual behaviour
Photo viewer show old image.

### Server configuration

**Operating system**:Debian 10

**Web server:** Nginx 1.17.1

**Database:** MariaDB 10.3

**PHP version:** 7.3.11

**Nextcloud version:** 16.0.6

**Updated from an older Nextcloud/ownCloud or fresh install:** Updated

**Where did you install Nextcloud from:** Download from official site

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.2.0
  - activity: 2.9.1
  - bruteforcesettings: 1.4.0
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - dav: 1.9.2
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_external: 1.7.0
  - files_pdfviewer: 1.5.0
  - files_rightclick: 0.15.1
  - files_sharing: 1.8.0
  - files_texteditor: 2.8.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - gallery: 18.3.0
  - groupfolders: 4.1.4
  - keeweb: 0.5.1
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - nextcloud_announcements: 1.5.0
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - password_policy: 1.6.0
  - privacy: 1.0.0
  - provisioning_api: 1.6.0
  - recommendations: 0.4.0
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - support: 1.0.0
  - survey_client: 1.4.0
  - systemtags: 1.6.0
  - theming: 1.7.0
  - twofactor_backupcodes: 1.5.0
  - updatenotification: 1.6.0
  - viewer: 1.2.0
  - workflowengine: 1.6.0
Disabled:
  - admin_audit
  - encryption
  - files_trashbin
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "hifiphile.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.6.1",
        "overwrite.cli.url": "https:\/\/removed_url.com",
        "installed": true,
        "maintenance": false,
        "theme": "",
        "loglevel": 2,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "mysql.utf8mb4": true,
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    },
    "apps": {
        "accessibility": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": ""
        },
        "activity": {
            "enabled": "yes",
            "installed_version": "2.9.1",
            "types": "filesystem"
        },
        "backgroundjob": {
            "lastjob": "6"
        },
        "bruteforcesettings": {
            "enabled": "yes",
            "installed_version": "1.4.0",
            "types": ""
        },
        "cloud_federation_api": {
            "enabled": "yes",
            "installed_version": "0.2.0",
            "types": "filesystem"
        },
        "comments": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "logging"
        },
        "core": {
            "default_encryption_module": "OC_DEFAULT_MODULE",
            "installedat": "1544477782.9284",
            "lastcron": "1573744152",
            "lastupdateResult": "[]",
            "lastupdatedat": "1573743375",
            "oc.integritycheck.checker": "[]",
            "public_files": "files_sharing\/public.php",
            "public_webdav": "dav\/appinfo\/v1\/publicwebdav.php",
            "scss.variables": "d40f8949538acad1fff0a11237ad2607",
            "theming.variables": "d40f8949538acad1fff0a11237ad2607",
            "updater.secret.created": "1573736281",
            "vendor": "nextcloud"
        },
        "dav": {
            "buildCalendarSearchIndex": "yes",
            "chunks_migrated": "1",
            "enabled": "yes",
            "installed_version": "1.9.2",
            "regeneratedBirthdayCalendarsForYearFix": "yes",
            "types": "filesystem"
        },
        "encryption": {
            "masterKeyId": "master_797e07fe",
            "publicShareKeyId": "pubShare_797e07fe",
            "recoveryKeyId": "recoveryKey_797e07fe"
        },
        "federatedfilesharing": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": ""
        },
        "federation": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "authentication"
        },
        "files": {
            "cronjob_scan_files": "500",
            "enabled": "yes",
            "installed_version": "1.11.0",
            "types": "filesystem"
        },
        "files_external": {
            "enabled": "yes",
            "installed_version": "1.7.0",
            "types": "filesystem"
        },
        "files_pdfviewer": {
            "enabled": "yes",
            "installed_version": "1.5.0",
            "types": ""
        },
        "files_rightclick": {
            "enabled": "yes",
            "installed_version": "0.15.1",
            "types": ""
        },
        "files_sharing": {
            "enabled": "yes",
            "installed_version": "1.8.0",
            "types": "filesystem"
        },
        "files_texteditor": {
            "enabled": "yes",
            "installed_version": "2.8.0",
            "types": ""
        },
        "files_trashbin": {
            "enabled": "no",
            "installed_version": "1.6.0",
            "types": "filesystem,dav"
        },
        "files_versions": {
            "enabled": "yes",
            "installed_version": "1.9.0",
            "types": "filesystem,dav"
        },
        "files_videoplayer": {
            "enabled": "yes",
            "installed_version": "1.5.0",
            "types": ""
        },
        "firstrunwizard": {
            "enabled": "yes",
            "installed_version": "2.5.0",
            "types": "logging"
        },
        "gallery": {
            "enabled": "yes",
            "installed_version": "18.3.0",
            "types": ""
        },
        "groupfolders": {
            "enabled": "yes",
            "installed_version": "4.1.4",
            "types": "filesystem,dav"
        },
        "keeweb": {
            "enabled": "yes",
            "installed_version": "0.5.1",
            "types": ""
        },
        "logreader": {
            "enabled": "yes",
            "installed_version": "2.1.0",
            "levels": "10111",
            "types": ""
        },
        "lookup_server_connector": {
            "enabled": "yes",
            "installed_version": "1.4.0",
            "types": "authentication"
        },
        "nextcloud_announcements": {
            "enabled": "yes",
            "installed_version": "1.5.0",
            "pub_date": "Thu, 24 Oct 2019 00:00:00 +0200",
            "types": "logging"
        },
        "notifications": {
            "enabled": "yes",
            "installed_version": "2.4.1",
            "types": "logging"
        },
        "oauth2": {
            "enabled": "yes",
            "installed_version": "1.4.2",
            "types": "authentication"
        },
        "password_policy": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": ""
        },
        "privacy": {
            "enabled": "yes",
            "installed_version": "1.0.0",
            "types": ""
        },
        "provisioning_api": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "prevent_group_restriction"
        },
        "recommendations": {
            "enabled": "yes",
            "installed_version": "0.4.0",
            "types": ""
        },
        "serverinfo": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": ""
        },
        "sharebymail": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "filesystem"
        },
        "support": {
            "enabled": "yes",
            "installed_version": "1.0.0",
            "types": ""
        },
        "survey_client": {
            "enabled": "yes",
            "installed_version": "1.4.0",
            "types": ""
        },
        "systemtags": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "logging"
        },
        "theming": {
            "backgroundMime": "image\/jpeg",
            "cachebuster": "12",
            "color": "#003C70",
            "enabled": "yes",
            "installed_version": "1.7.0",
            "logoMime": "image\/png",
            "logoheaderMime": "image\/png",
            "name": "HiFiPhile Master Storage",
            "slogan": "***REMOVED SENSITIVE VALUE***",
            "types": "logging",
            "url": "***REMOVED SENSITIVE VALUE***"
        },
        "twofactor_backupcodes": {
            "enabled": "yes",
            "installed_version": "1.5.0",
            "types": ""
        },
        "updatenotification": {
            "bruteforcesettings": "1.4.0",
            "core": "16.0.4.1",
            "enabled": "yes",
            "files_rightclick": "0.15.1",
            "groupfolders": "4.1.4",
            "installed_version": "1.6.0",
            "types": "",
            "update_check_errors": "0"
        },
        "viewer": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": ""
        },
        "workflowengine": {
            "enabled": "yes",
            "installed_version": "1.6.0",
            "types": "filesystem"
        }
    }
}

```
</details>

**Are you using external storage, if yes which one:** local

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
With access to your command line run e.g.:
sudo -u www-data php occ ldap:show-config
from within your Nextcloud installation folder

Without access to your command line download the data/owncloud.db to your local
computer or access your SQL server remotely and run the select query:
SELECT * FROM `oc_appconfig` WHERE `appid` = 'user_ldap';


Eventually replace sensitive data as the name/IP-address of your LDAP server or groups.
```
</details>

### Client configuration
**Browser:** Chrome 78

**Operating system:** Windows 10

#### Web server error log
<details>
<summary>Web server error log</summary>

```
No error in log

```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
No log generated while viewing the image.
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
Insert your browser log here, this could for example include:

a) The javascript console log
b) The network log
c) ...
```
</details>

**Comment by kesselb** [comment]
Are you updating the files via webdav?

**Comment by HiFiPhile** [comment]
Nextcloud server is running on a VM, images are generated by the host and share by a read-only NFS export.
I've also observed that the modified time is updated only if I've clicked the file, if I refresh the page modified time won't update.

**Comment by kesselb** [comment]
Probably a `occ files:scan` will help. The cleaner solution would be to update the files via webdav (mount the folder via webdav, write the files).

**Comment by HiFiPhile** [comment]
I'll try `occ files:scan`, is this a limitation of external storage ?

**Comment by kesselb** [comment]
https://docs.nextcloud.com/server/17/admin_manual/configuration_files/external_storage_configuration_gui.html#adding-files-to-external-storages

**Comment by HiFiPhile** [comment]
I've tried files:scan, but the issue persist. The modified time has been updated, but the old image still shown. I think it's better to check if the preview was generated before the last modification, in this case a new preview should be generated.
```
sudo -u www-data php occ files:scan --all
Starting scan for user 1 out of 2 ()
Starting scan for user 2 out of 2 ()
+---------+--------+--------------+
| Folders | Files  | Elapsed time |
+---------+--------+--------------+
| 43505   | 241740 | 00:18:07     |
+---------+--------+--------------+
```

**Comment by kesselb** [comment]
```
Index: lib/private/Preview/Generator.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lib/private/Preview/Generator.php	(revision 15f00f01263aa00d7722bf5618cb062170d5830e)
+++ lib/private/Preview/Generator.php	(date 1573749517794)
@@ -149,6 +149,9 @@
 		try {
 			try {
 				$preview = $this->getCachedPreview($previewFolder, $width, $height, $crop, $maxPreview->getMimeType(), $previewVersion);
+				if ($preview->getMTime() < $file->getMTime()) {
+					throw new NotFoundException('MTime preview is older then MTime file');
+				}
 			} catch (NotFoundException $e) {
 				$preview = $this->generatePreview($previewFolder, $maxPreview, $width, $height, $crop, $maxWidth, $maxHeight, $previewVersion);
 			}
```

Sounds good to me but could be a performance nightmare. Does it work for you as expected if you modify the code like above?

**Comment by HiFiPhile** [comment]
I tried this modification and reloaded nginx&php-fpm services but the problem still persist.
Then I tried to use only generatePreview to force preview generation but nothing happens.
```
// Try to get a cached preview. Else generate (and store) one
try {
	$preview = $this->generatePreview($previewFolder, $maxPreview, $width, $height, $crop, $maxWidth, $maxHeight);
	//try {
	//	$preview = $this->getCachedPreview($previewFolder, $width, $height, $crop, $maxPreview->getMimeType());
	//	if ($preview->getMTime() < $file->getMTime()) {
	//		throw new NotFoundException('MTime preview is older then MTime file');
	//	}
	//} catch (NotFoundException $e) {
	//	$preview = $this->generatePreview($previewFolder, $maxPreview, $width, $height, $crop, $maxWidth, $maxHeight);
	//}
} catch (\InvalidArgumentException $e) {
	throw new NotFoundException();
}
```
I've also cleared browser cache and used different browsers.

**Comment by kesselb** [comment]
cc @rullzer works as intended or bug?

**Comment by kesselb** [comment]
Let me reopen this one until we figured out if expected behaviour or not. It's probably easier to upload the files via curl: https://docs.nextcloud.com/server/latest/developer_manual/client_apis/WebDAV/basic.html

**Comment by szaimen** [comment]
Is this Issue still valid? If not, please close this issue. Thanks! :)

**Comment by HiFiPhile** [comment]
This issue is still valid on 20.0.8

**Comment by gennaios** [comment]
Is it the case that for modification of any  file, e.g., image, PDF, or other, the preview is not updated? There is this issue referring to markdown https://github.com/nextcloud/server/issues/24206.

I have not tested PDFs, but in my case, I have a preview generator for EPUBs, and the image is not updated on file change. I'd guess such may be the case not just for this issue but any file type.

**Comment by skjnldsv** [comment]
Cannot reproduce. tried that extensively while working on the new image editor.
Any PUT on the webdav endpoint properly cleans old previews

**Comment by WGandy** [comment]
I'm experiencing this issue on 25.03 with photos that have been changed. Interestingly, I have two browser pages open to the same folder of files and the one that was open while the file was changed (by a nextcloud client) doesn't show the  update in the "in browser" preview but the newly opened browser page does show the change (annotation added to the image with MacOS Preview app).  I've tried browsing to a different folder and then back but the preview image doesn't update. Also, the thumbnail DOES show the change. It seems to be a browser caching issue but still represents a usability issue for Nextcloud.
